### ISSUE TITLE
ErrorException in journals.twig

### ISSUE THREAD

**Issue by AxisNL** [JC5/firefly-iii#247]
Hmm.. I had some issues with accounts not showing any errors (got a whoops). I upgraded using a git upgrade and did a php artisan migrate, and now the accounts show transactions again. Nice.

However, no I can no longer show any transfers anymore (url /transactions/transfers), I get a whoops and this error in my mailbox:

Firefly III ran into an error: An exception has been thrown during the rendering of a template ("Trying to get property of non-object") in "list.journals" at line 53.

The error was of type "ErrorException".

The error occured on/at: Mon, 18 Apr 2016 14:28:15 +0000.

This error occured in file /home/domains/hongens/firefly/resources/views/list/journals.twig on line 53 with code 0.

The error was encountered by user #1, angelo@hongens.nl.

The IP address related to this error is: 192.168.128.14

The full stacktrace is below. If you think this is a bug in Firefly III, you can forward this message to thegrumpydictator@gmail.com. This can help fix the bug you just encountered.

If you prefer, you can also open a new issue on Github.

The full stacktrace is below: <nothing shown>

Any ideas?

**Comment by AxisNL** [comment]
Enabled debug for the stack trace:

Error

This error occured in file /home/domains/hongens/firefly/resources/views/list/journals.twig on line 53 with code 0.

Stack trace

#0 /home/domains/hongens/firefly/vendor/rcrowe/twigbridge/src/Engine/Twig.php(92): TwigBridge\Engine\Twig->handleTwigError(Object(Twig_Error_Runtime))
#1 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(14849): TwigBridge\Engine\Twig->get('/home/domains/h...', Array)
#2 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(14837): Illuminate\View\View->getContents()
#3 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(14821): Illuminate\View\View->renderContents()
#4 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(16040): Illuminate\View\View->render()
#5 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(15539): Illuminate\Http\Response->setContent(Object(Illuminate\View\View))
#6 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(8234): Symfony\Component\HttpFoundation\Response->__construct(Object(Illuminate\View\View))
#7 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9310): Illuminate\Routing\Router->prepareResponse(Object(Illuminate\Http\Request), Object(Illuminate\View\View))
#8 [internal function]: Illuminate\Routing\ControllerDispatcher->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#9 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(52): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#10 [internal function]: Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#11 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9787): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#12 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9311): Illuminate\Pipeline\Pipeline->then(Object(Closure))
#13 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9298): Illuminate\Routing\ControllerDispatcher->callWithinStack(Object(FireflyIII\Http\Controllers\TransactionController), Object(Illuminate\Routing\Route), Object(Illuminate\Http\Request), 'index')
#14 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(8385): Illuminate\Routing\ControllerDispatcher->dispatch(Object(Illuminate\Routing\Route), Object(Illuminate\Http\Request), 'FireflyIII\\Http...', 'index')
#15 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(8372): Illuminate\Routing\Route->runController(Object(Illuminate\Http\Request))
#16 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(8086): Illuminate\Routing\Route->run(Object(Illuminate\Http\Request))
#17 [internal function]: Illuminate\Routing\Router->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#18 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(52): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#19 /home/domains/hongens/firefly/app/Http/Middleware/Binder.php(45): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#20 [internal function]: FireflyIII\Http\Middleware\Binder->handle(Object(Illuminate\Http\Request), Object(Closure))
#21 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9797): call_user_func_array(Array, Array)
#22 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#23 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(32): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#24 /home/domains/hongens/firefly/app/Http/Middleware/Range.php(90): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#25 [internal function]: FireflyIII\Http\Middleware\Range->handle(Object(Illuminate\Http\Request), Object(Closure))
#26 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9797): call_user_func_array(Array, Array)
#27 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#28 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(32): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#29 /home/domains/hongens/firefly/app/Http/Middleware/IsConfirmed.php(59): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#30 [internal function]: FireflyIII\Http\Middleware\IsConfirmed->handle(Object(Illuminate\Http\Request), Object(Closure))
#31 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9797): call_user_func_array(Array, Array)
#32 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#33 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(32): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#34 /home/domains/hongens/firefly/app/Http/Middleware/AuthenticateTwoFactor.php(63): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#35 [internal function]: FireflyIII\Http\Middleware\AuthenticateTwoFactor->handle(Object(Illuminate\Http\Request), Object(Closure))
#36 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9797): call_user_func_array(Array, Array)
#37 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#38 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(32): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#39 /home/domains/hongens/firefly/app/Http/Middleware/Authenticate.php(44): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#40 [internal function]: FireflyIII\Http\Middleware\Authenticate->handle(Object(Illuminate\Http\Request), Object(Closure))
#41 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9797): call_user_func_array(Array, Array)
#42 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#43 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(32): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#44 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(3165): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#45 [internal function]: Illuminate\Foundation\Http\Middleware\VerifyCsrfToken->handle(Object(Illuminate\Http\Request), Object(Closure))
#46 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9797): call_user_func_array(Array, Array)
#47 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#48 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(32): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#49 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(13221): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#50 [internal function]: Illuminate\View\Middleware\ShareErrorsFromSession->handle(Object(Illuminate\Http\Request), Object(Closure))
#51 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9797): call_user_func_array(Array, Array)
#52 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#53 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(32): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#54 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(11781): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#55 [internal function]: Illuminate\Session\Middleware\StartSession->handle(Object(Illuminate\Http\Request), Object(Closure))
#56 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9797): call_user_func_array(Array, Array)
#57 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#58 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(32): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#59 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(12960): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#60 [internal function]: Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse->handle(Object(Illuminate\Http\Request), Object(Closure))
#61 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9797): call_user_func_array(Array, Array)
#62 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#63 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(32): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#64 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(12897): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#65 [internal function]: Illuminate\Cookie\Middleware\EncryptCookies->handle(Object(Illuminate\Http\Request), Object(Closure))
#66 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9797): call_user_func_array(Array, Array)
#67 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#68 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(32): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#69 [internal function]: Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#70 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9787): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#71 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(8087): Illuminate\Pipeline\Pipeline->then(Object(Closure))
#72 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(8078): Illuminate\Routing\Router->runRouteWithinStack(Object(Illuminate\Routing\Route), Object(Illuminate\Http\Request))
#73 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(8068): Illuminate\Routing\Router->dispatchToRoute(Object(Illuminate\Http\Request))
#74 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(2393): Illuminate\Routing\Router->dispatch(Object(Illuminate\Http\Request))
#75 [internal function]: Illuminate\Foundation\Http\Kernel->Illuminate\Foundation\Http\{closure}(Object(Illuminate\Http\Request))
#76 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(52): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#77 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(3226): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#78 [internal function]: Illuminate\Foundation\Http\Middleware\CheckForMaintenanceMode->handle(Object(Illuminate\Http\Request), Object(Closure))
#79 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9797): call_user_func_array(Array, Array)
#80 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#81 /home/domains/hongens/firefly/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(32): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#82 [internal function]: Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#83 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(9787): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#84 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(2340): Illuminate\Pipeline\Pipeline->then(Object(Closure))
#85 /home/domains/hongens/firefly/bootstrap/cache/compiled.php(2324): Illuminate\Foundation\Http\Kernel->sendRequestThroughRouter(Object(Illuminate\Http\Request))
#86 /home/domains/hongens/firefly/public/index.php(53): Illuminate\Foundation\Http\Kernel->handle(Object(Illuminate\Http\Request))
#87 {main}

**Comment by JC5** [comment]
I have no idea. That line has nothing special at all. Or rather, the code there is so basic that if that is broken, unfortunately, I have no idea what is going on.

During the loop that particular line is part of, this happens:

```
{% for journal in journals %}
...
{% endfor %}
```

Which is a pretty basic for-loop. Iterating over a collection of journals.

Line 53 in turn, is this: ``{{ journal|formatJournal }}``.

I would expect an error to appear in the method "formatJournal", to which this is referring. But the first property Firefly tries to access (remember the error: "Trying to get property of non-object") is the 'id', which is the most important property of the object (aka, the transaction). And if that one does not exist, something very strange is going on.

And I have no idea what it could be.

**Comment by AxisNL** [comment]
I did have some issues last month where I did a stupid import and I manually removed some stuff from the DB. I think I removed some transactions, but my guess is I left the DB in an inconsistent state. Perhaps these issues are the result of the leftovers.

For the short run: if I buy you a beer, will you help me get my DB back in a consistent state?

For the long run: Since I might not be the only one to ever have touched the database in a stupid way, I would really like a method to verify the database integrity. Perhaps a 'php artisan verifydb' command?

**Comment by JC5** [comment]
Hah. Sounds good!

The easiest way to verify if the database is in a consistent state, is visiting the home page.

See [here](https://github.com/JC5/firefly-iii/blob/master/app/Http/Controllers/HomeController.php#L121) and [here](https://github.com/JC5/firefly-iii/blob/master/app/Repositories/Account/AccountRepository.php#L559) and [here](https://github.com/JC5/firefly-iii/blob/master/resources/lang/en_US/firefly.php#L35).

You should get a big fat warning if there is anything strange going on.

Having said that, I'll think of some queries which should show you inconsistencies. After that, I'll indeed build some verification tool for the command line, or to be accessed from some URL.

Expect a reply today or tomorrow!

**Comment by JC5** [comment]
This query should return any deleted transaction journals that have transactions that are NOT deleted:

```select transaction_journals.id as journal_id,transaction_journals.deleted_at as journal_deleted, transactions.id as transaction_id, transactions.deleted_at as transaction_deleted_at from transaction_journals left join transactions ON transactions.transaction_journal_id = transaction_journals.id where transaction_journals.deleted_at is not null AND transactions.deleted_at is null and transactions.id is not null``` 

To fix this, open all transactions (``transaction_id``) and change the ``deleted_at`` field to a time in the past.

The other way around: deleted transactions that are connected to a not deleted journal.

```select transactions.id as transaction_id, transactions.deleted_at as transaction_deleted, transaction_journals.id as journal_id, transaction_journals.deleted_at  from transactions left join transaction_journals ON transactions.transaction_journal_id = transaction_journals.id where transactions.deleted_at is not null and transaction_journals.deleted_at is null```

Should report on deleted accounts that still have not deleted transactions or journals attached to them.

```select accounts.id as account_id, accounts.deleted_at as account_deleted_at, transactions.id as transaction_id, transactions.deleted_at as transaction_deleted_at,
transaction_journals.id as journal_id, transaction_journals.deleted_at as journal_deleted_at from accounts left join transactions ON transactions.account_id = accounts.id left join transaction_journals ON transaction_journals.id = transactions.transaction_journal_id where accounts.deleted_at is not null and transactions.id is not null and (transactions.deleted_at is null OR transaction_journals.deleted_at is null)```

**Comment by AxisNL** [comment]
Thanks for the elaborate answer! I hope this will help future peeps as well. For me, all queries returned 0 records.. I'll look a bit further when I have time..

**Comment by JC5** [comment]
If it is not, try the following. The current version of FF allows you to set a page length. Set it to 25.

Try to find a page using the URL as follows (``/transactions/withdrawal?page=x``) and report here when Firefly breaks.

A new transaction should not break. So you should be able to set the page size to 1, create a new transaction and view page  ``/transactions/withdrawal?page=x``. Other pages may break again.

It should be fairly easy to deduce which transactions it's about. Then we can try and find out what's happening.

**Comment by JC5** [comment]
Welp, I currently have the same problem in my test database. Will report back.

**Comment by JC5** [comment]
I think the firefly:verify command should catch most of these problems. It will be included in 3.9.0

It can be run using:

``php artisan firefly:verify``

Please let me know if the problem is reported.
